In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import os
os.environ["OMP_NUM_THREADS"] = "1"
import pprint

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
df_market_data.dtypes

price_change_percentage_24h     float64
price_change_percentage_7d      float64
price_change_percentage_14d     float64
price_change_percentage_30d     float64
price_change_percentage_60d     float64
price_change_percentage_200d    float64
price_change_percentage_1y      float64
dtype: object

In [6]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data[
                                                    ["price_change_percentage_24h",
                                                     "price_change_percentage_7d", 
                                                     "price_change_percentage_14d", 
                                                     "price_change_percentage_30d",
                                                     "price_change_percentage_60d", 
                                                     "price_change_percentage_200d", 
                                                     "price_change_percentage_1y"
                                                     ]])

In [7]:
# Create a DataFrame with the scaled data

df_market_data_scaled = pd.DataFrame(market_data_scaled, columns = [
                                                     "price_change_percentage_24h",
                                                     "price_change_percentage_7d", 
                                                     "price_change_percentage_14d", 
                                                     "price_change_percentage_30d",
                                                     "price_change_percentage_60d", 
                                                     "price_change_percentage_200d", 
                                                     "price_change_percentage_1y"
                                                     ])
# Copy the crypto names from the original DataFrame

df_market_data_scaled['coin_id'] = df_market_data.index
# Set the coin_id column as index

df_market_data_scaled.set_index('coin_id', inplace=True)
# Display the scaled DataFrame
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [8]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 11))

In [9]:
# Create an empty list to store the inertia values
inertia = []



In [10]:
# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans = KMeans(n_clusters=k, random_state=0)
    
    # 2. Fit the model to the data using df_market_data_scaled
    kmeans.fit(df_market_data_scaled)
    
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(kmeans.inertia_)

# Display the inertia values
print("Inertia values for k from 1 to 11:", inertia)

Inertia values for k from 1 to 11: [287.0, 198.5718180979813, 145.63574772192862, 79.02243535120978, 66.21778006000338, 59.13907002537009, 48.40631449041362, 38.99358798927969, 34.40059545400235, 28.57360256096785]


In [11]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k_values, "inertia": inertia}

# Check the dictionary
pprint.pprint(elbow_data)



{'inertia': [287.0,
             198.5718180979813,
             145.63574772192862,
             79.02243535120978,
             66.21778006000338,
             59.13907002537009,
             48.40631449041362,
             38.99358798927969,
             34.40059545400235,
             28.57360256096785],
 'k': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}


In [12]:

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
# Print the DataFrame
print(df_elbow)

    k     inertia
0   1  287.000000
1   2  198.571818
2   3  145.635748
3   4   79.022435
4   5   66.217780
5   6   59.139070
6   7   48.406314
7   8   38.993588
8   9   34.400595
9  10   28.573603


In [13]:
df_elbow.hvplot.line(x="k", y="inertia", title="Elbow Curve", xlabel="Number of clusters (k)", ylabel="Inertia", line_width=2)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for K is 4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [14]:
# Initialize the K-Means model using the best value for k
k=4
kmeans = KMeans(n_clusters=k, random_state=0)

In [15]:
# Fit the K-Means model using the scaled DataFrame
kmeans.fit(df_market_data_scaled)

KMeans(n_clusters=4, random_state=0)

In [16]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
clusters = kmeans.predict(df_market_data_scaled)

#  Print the resulting array of cluster values
print("Cluster assignments for each cryptocurrency:")
print(clusters)



Cluster assignments for each cryptocurrency:
[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [17]:
# Create a copy of the scaled DataFrame
df_market_data_scaled_copy = df_market_data_scaled.copy()



# Display the first few rows of the copied DataFrame with clusters
df_market_data_scaled_copy.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

In [18]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters

df_market_data_scaled_copy['cluster'] = clusters

# Display the first few rows of the copied DataFrame with clusters
df_market_data_scaled_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster  
coin_id                                            
bitcoin                        -0.251637        2  
ethereum                       -0.199352        2  
tether                         -0.282061        0  
ripple                         -0.295546        0  
bitcoin-cash                   -0.270317        2

In [19]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.


In [20]:
# Create a scatter plot using hvPlot
scatter_plot = df_market_data_scaled_copy.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d", 
    c='cluster', 
    cmap='viridis', 
    hover_cols=['coin_id'],
    title="Cryptocurrency Clusters",
    width=800, 
    height=400
    )

# Display the scatter plot
scatter_plot

:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cluster,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [21]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [22]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_data = pca.fit_transform(df_market_data_scaled)

# View the scaled PCA data
df_pca = pd.DataFrame(pca_data, columns=["PC1", "PC2", "PC3"], index=df_market_data_scaled.index)
print(df_pca.head())

                   PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715


In [23]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_


In [24]:
print("Explained variance for each principal component:", explained_variance)

Explained variance for each principal component: [0.3719856  0.34700813 0.17603793]


In [25]:
total_explained_variance = explained_variance.sum()
print("Total explained variance of the three principal components:", total_explained_variance)

Total explained variance of the three principal components: 0.8950316570309841


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:Total Explained Variance: 0.8950 (89.50%)** 

In [26]:
# Create a new DataFrame with the PCA data.
df_pca = pd.DataFrame(pca_data, columns=["PC1", "PC2", "PC3"])

# Copy the crypto names from the original scaled DataFrame

df_pca['coin_id'] = df_market_data_scaled.index
# Set the coin_id column as index
df_pca.set_index('coin_id', inplace=True)


# Display the scaled PCA DataFrame
print(df_pca.head())


                   PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715


---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [27]:
# Create a list with the number of k-values from 1 to 11
k_values_pca = list(range(1, 11))

In [28]:
# Create an empty list to store the inertia values
inertia_pca = []
 
# Create a for loop to compute the inertia with each possible value of k
for k in k_values_pca:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
     kmeans_pca = KMeans(n_clusters=k, random_state=0)
# 2. Fit the model to the data using `df_market_data_pca`
     kmeans_pca.fit(df_pca)
# 3. Append the model.inertia_ to the inertia list
     inertia_pca.append(kmeans_pca.inertia_)

In [29]:
# Display the inertia values
print("Inertia values for k from 1 to 11 using PCA data:", inertia_pca)

Inertia values for k from 1 to 11 using PCA data: [256.87408556789256, 168.81189630685228, 115.91807663320066, 49.66549665179738, 41.35731380312857, 30.79664178704162, 25.915456910456157, 21.890936274633237, 17.720698854473678, 13.838698708864579]


In [30]:

# Print the inertia values
pprint.pprint(inertia_pca)

[256.87408556789256,
 168.81189630685228,
 115.91807663320066,
 49.66549665179738,
 41.35731380312857,
 30.79664178704162,
 25.915456910456157,
 21.890936274633237,
 17.720698854473678,
 13.838698708864579]


In [31]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k_values_pca, "inertia": inertia_pca}

# Print the dictionary
pprint.pprint(elbow_data_pca)

{'inertia': [256.87408556789256,
             168.81189630685228,
             115.91807663320066,
             49.66549665179738,
             41.35731380312857,
             30.79664178704162,
             25.915456910456157,
             21.890936274633237,
             17.720698854473678,
             13.838698708864579],
 'k': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}


In [32]:
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)

# Print the DataFrame
df_elbow_pca

k     inertia
0   1  256.874086
1   2  168.811896
2   3  115.918077
3   4   49.665497
4   5   41.357314
5   6   30.796642
6   7   25.915457
7   8   21.890936
8   9   17.720699
9  10   13.838699

In [33]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
# Plot a line chart to visualize the Elbow curve
df_elbow_pca.hvplot.line(
    x="k", y="inertia", title="Elbow Curve (PCA Data)", 
    xlabel="Number of clusters (k)", ylabel="Inertia", line_width=2
)


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** k=4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** no difference

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [34]:
# Initialize the K-Means model using the best value for k
best_k = 4   
kmeans_pca = KMeans(n_clusters=best_k, random_state=0)

In [35]:
# Fit the K-Means model using the PCA data
kmeans_pca.fit(df_pca)

KMeans(n_clusters=4, random_state=0)

In [36]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
clusters_pca = kmeans_pca.predict(df_pca)


In [37]:

# Print the resulting array of cluster values.
print("Cluster assignments for each cryptocurrency using PCA:")
print(clusters_pca)

Cluster assignments for each cryptocurrency using PCA:
[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [38]:
# Create a copy of the scaled PCA DataFrame
df_pca_copy = df_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_pca_copy['cluster'] = clusters_pca


In [39]:

# Display the copy of the scaled PCA DataFrame
print(df_pca_copy.head())

                   PC1       PC2       PC3  cluster
coin_id                                            
bitcoin      -0.600667  0.842760  0.461595        2
ethereum     -0.458261  0.458466  0.952877        2
tether       -0.433070 -0.168126 -0.641752        0
ripple       -0.471835 -0.222660 -0.479053        0
bitcoin-cash -1.157800  2.041209  1.859715        2


In [40]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

df_pca_copy.hvplot.scatter(
    x="PC1", y="PC2", 
    color="cluster", 
    hover_cols=['coin_id'], 
    title="Cryptocurrency Clusters using PCA Data",
    width=800, height=400
)


:Scatter   [PC1]   (PC2,cluster,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [41]:


# Create a DataFrame for the Elbow curve with original scaled data
elbow_data_original = pd.DataFrame({"k": k_values, "inertia": inertia})

# Create a DataFrame for the Elbow curve with PCA-transformed data
elbow_data_pca = pd.DataFrame({"k": k_values, "inertia": inertia_pca})

# Plot the Elbow curve for the original scaled data
elbow_curve_original = elbow_data_original.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve (Original Scaled Data)", 
    xlabel="Number of clusters (k)", 
    ylabel="Inertia",
    color="blue", 
    line_width=2, 
    width=400, 
    height=300
)

# Plot the Elbow curve for the PCA-transformed data
elbow_curve_pca = elbow_data_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve (PCA Data)", 
    xlabel="Number of clusters (k)", 
    ylabel="Inertia",
    color="green", 
    line_width=2, 
    width=400, 
    height=300
)

# Combine the two Elbow curves in a composite plot
composite_plot = (elbow_curve_original + elbow_curve_pca).cols(2)

# Display the composite plot
composite_plot


:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [42]:
# Composite plot to contrast the clusters
# # 1. Scatter plot for clusters using the original scaled data
scatter_original = df_market_data_scaled_copy.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d", 
    color="cluster", 
    hover_cols=['coin_id'], 
    title="Clusters (Original Scaled Data)",
    width=400, 
    height=400
   
)

# 2. Scatter plot for clusters using the PCA-transformed data
scatter_pca = df_pca_copy.hvplot.scatter(
    x="PC1", 
    y="PC2", 
    color="cluster", 
    hover_cols=['coin_id'], 
    title="Clusters (PCA Data)",
    width=400, 
    height=400
)

# 3. Composite plot to contrast the clusters
composite_clusters_plot = (scatter_original + scatter_pca).cols(2)

# Display the composite plot
composite_clusters_plot


:Layout
   .Scatter.I  :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cluster,coin_id)
   .Scatter.II :Scatter   [PC1]   (PC2,cluster,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?



 * **Answer:** 

PCA reduces the number of features while retaining the most important information. By focusing on key patterns, it simplifies the data, allowing K-Means to find clusters more efficiently.<br>
Clustering on PCA-transformed data is **_clearer, more distinct clusters_** compared to using the full set of features<br>

By reducing dimensions, PCA has **_separated clusters more effectively_**.<br> It eliminated the redundant/correlated features, allowing the K-Means algorithm to better distinguish between clusters.<br>
The clusters appear more distinct in lower dimensions, improving interpretability. The Elbow curve for PCA data did not suggest a different optimal number of clusters compared to the full feature set.


The composite plots of clusters, make it easier to spot visually distinct clusters and interpret how the data points are grouped.